### ISSUE TITLE
[BUG] The unit test `sorted partitioned write` fails on Spark 3.4

### ISSUE THREAD

**Issue by HaoYang670** [NVIDIA/spark-rapids#7018]
**Describe the bug**
```
- sorted partitioned write *** FAILED ***
  Expected "000[0000001]", but got "000[9942054]" (ParquetWriterSuite.scala:99)
```
**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.
Run
```
mvn clean package -DwildcardSuites=com.nvidia.spark.rapids.ParquetWriterSuite -Dbuildver=340
```
**Expected behavior**
The test should pass.

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

**Additional context**

**Comment by HaoYang670** [comment]
This failure is very similar to https://github.com/apache/spark/pull/38356. cc @jlowe is it a bug in Spark?

**Comment by jlowe** [comment]
If the test fails on Spark 3.4.0 even on CPU then it may be the same issue as you linked.  If it regularly passes on Spark 3.4.0 on the CPU but always fails on the GPU then it's likely a problem in the plugin.

**Comment by HaoYang670** [comment]
Thank you, @jlowe. The test can pass on Spark 3.4.0 on the CPU, which indicates that the bug is in our plugin.

**Comment by HaoYang670** [comment]
By the way. The Spark currently **does** resort in `partitionBy`, which is different from our plugin.
Tested on Spark3.3.0
```scala
scala> val df = sc.parallelize((1L to 10000000L)).map{i => ("a", f"$i%010d", i)}.toDF("partkey", "val", "val2")
df: org.apache.spark.sql.DataFrame = [partkey: string, val: string ... 1 more field]

scala> df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val", $"val2").write.mode("overwrite").partitionBy("partkey").parquet(path)
                                                                                
scala> spark.read.parquet(path).limit(10).show
+----------+-------+-------+
|       val|   val2|partkey|
+----------+-------+-------+
|0000000001|      1|      a|
|0008248673|8248673|      a|
|0000000002|      2|      a|
|0008248674|8248674|      a|
|0000000003|      3|      a|
|0008248675|8248675|      a|
|0000000004|      4|      a|
|0008248676|8248676|      a|
|0000000005|      5|      a|
|0008248677|8248677|      a|
+----------+-------+-------+


scala> df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val", $"val2").write.mode("overwrite").parquet(path)
                                                                                
scala> spark.read.parquet(path).limit(10).show
+-------+----------+----+
|partkey|       val|val2|
+-------+----------+----+
|      a|0000000001|   1|
|      a|0000000002|   2|
|      a|0000000003|   3|
|      a|0000000004|   4|
|      a|0000000005|   5|
|      a|0000000006|   6|
|      a|0000000007|   7|
|      a|0000000008|   8|
|      a|0000000009|   9|
|      a|0000000010|  10|
+-------+----------+----+
```

This could be a bug in Spark, but I am sure whether should we revert https://github.com/NVIDIA/spark-rapids/issues/3002 to match Spark's behaviour.

**Comment by HaoYang670** [comment]
Also, the test `sorted_partitioned_write` could fail if I set the conf `"spark.rapids.sql.format.parquet.write.enabled"` to `true`, is this a bug?

**Comment by HaoYang670** [comment]
Another related issue of spark: https://issues.apache.org/jira/browse/SPARK-40588, which will affect not only Spark 3.4 but also other Spark3.2+ versions

**Comment by jlowe** [comment]
We should not revert #3002.  That was a true bug where we were not propagating the sort order for an operation that does not modify the sort order.  Undoing that will likely introduce a problem in another scenario where Spark is currently doing the right thing across all versions.

In this case, doing the right thing should be fine, as that's doing a superset of the partition-key-only sort.  The data may not match exactly what the (broken) Spark behavior is, but I think that should be fine.  We can document the issue as a bug in some Spark versions we are not replicating.

As for the Spark 3.4 issue, examine the query plans for the CPU and GPU versions.  Is the CPU plan also (incorrectly) throwing away the full sort ahead of the partition-only sort and we're getting unlucky in the test, where it just happens to be accidentally sorted on the other columns when only sorting on the partition column?  Or do the CPU and GPU plans look semantically different, where the CPU plan is doing a sort across all three columns but the GPU plan is only sorting on the partition column (possibly implicitly in the partitioned write)?

**Comment by HaoYang670** [comment]
> We should not revert https://github.com/NVIDIA/spark-rapids/issues/3002. 

> The data may not match exactly what the (broken) Spark behavior is, but I think that should be fine. We can document the issue as a bug in some Spark versions we are not replicating.

Makes sense. Thank you @jlowe.

**Comment by HaoYang670** [comment]
For Spark 3.4, we get these results when running on CPU:
```
scala> val df = sc.parallelize((1L to 10000000L)).map{i => ("a", f"$i%010d", i)}.toDF("partkey", "val", "val2")
df: org.apache.spark.sql.DataFrame = [partkey: string, val: string ... 1 more field]

scala> val path = "~/working/spark/sort_partition.parquet"
path: String = ~/working/spark/sort_partition.parquet

scala> df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val", $"val2").write.mode("overwrite").partitionBy("partkey").parquet(path)
                                                                                
scala> spark.read.parquet(path).limit(10).show;
+----------+-------+-------+
|       val|   val2|partkey|
+----------+-------+-------+
|0000000001|      1|      a|
|0008283625|8283625|      a|
|0000000002|      2|      a|
|0008283626|8283626|      a|
|0000000003|      3|      a|
|0008283627|8283627|      a|
|0000000004|      4|      a|
|0008283628|8283628|      a|
|0000000005|      5|      a|
|0008283629|8283629|      a|
+----------+-------+-------+
```

And running the plugin, we get 
```
scala> spark.conf.set("spark.rapids.sql.format.parquet.enabled",false)
...
+----------+-------+-------+
|       val|   val2|partkey|
+----------+-------+-------+
|0009942054|9942054|      a|
|0009942055|9942055|      a|
|0009942056|9942056|      a|
|0009942057|9942057|      a|
|0009942058|9942058|      a|
|0009942059|9942059|      a|
|0009942060|9942060|      a|
|0009942061|9942061|      a|
|0009942062|9942062|      a|
|0009942063|9942063|      a|
+----------+-------+-------+

```

**Comment by HaoYang670** [comment]
It seems rows can be correctly sorted (I mean sort by the 3 columns and `partitionBy` won't introduce resorting) on GPU, but the start row is not  `|0009942054|9942054|      a|`. I am afraid this is because some data is spilled to disk.

When I reduce the size of the dataframe, I can get the expected result.
This is also run by GPU:
```
// half the dataset
scala> val df = sc.parallelize((1L to 5000000L)).map{i => ("a", f"$i%010d", i)}.toDF("partkey", "val", "val2")
...
+----------+----+-------+
|       val|val2|partkey|
+----------+----+-------+
|0000000001|   1|      a|
|0000000002|   2|      a|
|0000000003|   3|      a|
|0000000004|   4|      a|
|0000000005|   5|      a|
|0000000006|   6|      a|
|0000000007|   7|      a|
|0000000008|   8|      a|
|0000000009|   9|      a|
|0000000010|  10|      a|
+----------+----+-------+
```

**Comment by jlowe** [comment]
As I mentioned above, we need to examine the query plans on the CPU and GPU to see what's going on.  In both cases, I see the sort across multiple keys being thrown away for a sort on just the partition key.  Therefore both the CPU and GPU answers are "correct" regarding what was planned (but both are not correct regarding the original query's intent).  This is a problem with Spark's planning, not with the plugin.

The fact that the CPU shows the correct result is by chance, not by design.  This is most easily shown by slightly modifying the query so the sort values don't monotonically increase (i.e.: come out already in the desired sort order), e.g.:
```
// Run on the CPU
scala> val df = sc.parallelize((1L to 10000000L)).map{i => ("a", f"$i%010d", 10000000-i)}.toDF("partkey", "val", "val2")
df: org.apache.spark.sql.DataFrame = [partkey: string, val: string ... 1 more field]

// This is the intended order to be written
scala> df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val2").show(10)
+-------+----------+----+                                                       
|partkey|       val|val2|
+-------+----------+----+
|      a|0010000000|   0|
|      a|0009999999|   1|
|      a|0009999998|   2|
|      a|0009999997|   3|
|      a|0009999996|   4|
|      a|0009999995|   5|
|      a|0009999994|   6|
|      a|0009999993|   7|
|      a|0009999992|   8|
|      a|0009999991|   9|
+-------+----------+----+
only showing top 10 rows


scala> df.repartition(1, $"partkey").sortWithinPartitions($"partkey", $"val2").write.mode("overwrite").partitionBy("partkey").parquet(path)

// Notice the write drops the desired sort order, even on the CPU                                                                                
scala> spark.read.parquet(path).show(10)
+----------+-------+-------+
|       val|   val2|partkey|
+----------+-------+-------+
|0000000001|9999999|      a|
|0000000002|9999998|      a|
|0000000003|9999997|      a|
|0000000004|9999996|      a|
|0000000005|9999995|      a|
|0000000006|9999994|      a|
|0000000007|9999993|      a|
|0000000008|9999992|      a|
|0000000009|9999991|      a|
|0000000010|9999990|      a|
+----------+-------+-------+
only showing top 10 rows
```

TLDR; This is fallout from the Spark bug where sort order keys are getting dropped during partitioned writes.  Both the CPU and GPU outputs are "correct" with respect to how the query was planned.  The bug is in how the query was planned.

**Comment by nartal1** [comment]
@jlowe Is it okay if we disable this test for Spark-3.4.0 and revisit this once the bug is fixed in Apache Spark. We will keep the issue open and enable the test whenever the bug in Apache Spark is fixed. 
The reason I am requesting for this exception is we can enable unit tests in pre-merge(for Spark-3.4 if required) and in nightly builds. 
Currently 4 tests are failing for Spark-3.4 of which 3 tests will be fixed by this PR -  https://github.com/NVIDIA/spark-rapids/pull/7099.

**Comment by jlowe** [comment]
Yes, I'm totally OK with ignoring this test for Spark 3.4 as long as we point to the Spark issue in the code explaining why.
